File: bindings/swift/Tests/UfsecpTests/SmokeTest.swift
// ============================================================================
// UltrafastSecp256k1 — Swift Binding Smoke Test (Golden Vectors)
// ============================================================================
// Verifies FFI boundary correctness using deterministic known-answer tests.
// Runs in <2 seconds.
//
// Usage:
//   swift test --filter SmokeTest
// ============================================================================

import XCTest
@testable import Ufsecp

final class SmokeTest: XCTestCase {

    // ── Golden Vectors ──────────────────────────────────────────────────

    static let KNOWN_PRIVKEY = Data(hexString:
        "0000000000000000000000000000000000000000000000000000000000000001")!

    static let KNOWN_PUBKEY = Data(hexString:
        "0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798")!

    static let KNOWN_XONLY = Data(hexString:
        "79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798")!

    static let SHA256_EMPTY = Data(hexString:
        "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855")!

    let msg32 = Data(count: 32)
    let aux32 = Data(count: 32)

    var ctx: UfsecpContext!

    override func setUp() {
        super.setUp()
        ctx = try! UfsecpContext()
    }

    override func tearDown() {
        ctx = nil
        super.tearDown()
    }

    // ── Tests ───────────────────────────────────────────────────────────

    func testCtxCreateAbi() throws {
        let abi = ctx.abiVersion
        XCTAssertGreaterThanOrEqual(abi, 1, "ABI version < 1")
    }

    func testPubkeyCreateGolden() throws {
        let pub = try ctx.pubkeyCreate(privkey: Self.KNOWN_PRIVKEY)
        XCTAssertEqual(pub, Self.KNOWN_PUBKEY, "compressed pubkey mismatch")
    }

    func testPubkeyXonlyGolden() throws {
        let xonly = try ctx.pubkeyXonly(privkey: Self.KNOWN_PRIVKEY)
        XCTAssertEqual(xonly, Self.KNOWN_XONLY, "x-only pubkey mismatch")
    }

    func testEcdsaSignVerify() throws {
        let sig = try ctx.ecdsaSign(msgHash: msg32, privkey: Self.KNOWN_PRIVKEY)
        XCTAssertEqual(sig.count, 64, "sig length")

        let ok = try ctx.ecdsaVerify(msgHash: msg32, sig: sig, pubkey: Self.KNOWN_PUBKEY)
        XCTAssertTrue(ok, "valid sig rejected")

        // Mutated → fail
        var bad = sig
        bad[0] ^= 0x01
        let fail = try ctx.ecdsaVerify(msgHash: msg32, sig: bad, pubkey: Self.KNOWN_PUBKEY)
        XCTAssertFalse(fail, "mutated sig accepted")
    }

    func testSchnorrSignVerify() throws {
        let sig = try ctx.schnorrSign(msg: msg32, privkey: Self.KNOWN_PRIVKEY, auxRand: aux32)
        XCTAssertEqual(sig.count, 64, "schnorr sig length")

        let ok = try ctx.schnorrVerify(msg: msg32, sig: sig, pubkeyX: Self.KNOWN_XONLY)
        XCTAssertTrue(ok, "valid schnorr sig rejected")
    }

    func testEcdsaRecover() throws {
        let rec = try ctx.ecdsaSignRecoverable(msgHash: msg32, privkey: Self.KNOWN_PRIVKEY)
        let pub = try ctx.ecdsaRecover(msgHash: msg32, sig: rec.signature, recid: rec.recoveryId)
        XCTAssertEqual(pub, Self.KNOWN_PUBKEY, "recovered pubkey mismatch")
    }

    func testSha256Golden() throws {
        let digest = try ctx.sha256(Data())
        XCTAssertEqual(digest, Self.SHA256_EMPTY, "SHA-256 empty mismatch")
    }

    func testAddrP2wpkh() throws {
        let addr = try ctx.addrP2wpkh(pubkey: Self.KNOWN_PUBKEY, network: .mainnet)
        XCTAssertTrue(addr.hasPrefix("bc1q"), "Expected bc1q..., got \(addr)")
    }

    func testWifRoundtrip() throws {
        let wif = try ctx.wifEncode(privkey: Self.KNOWN_PRIVKEY, compressed: true, network: .mainnet)
        let decoded = try ctx.wifDecode(wif: wif)
        XCTAssertEqual(decoded.privkey, Self.KNOWN_PRIVKEY, "WIF privkey")
        XCTAssertTrue(decoded.compressed, "WIF compressed")
    }

    func testEcdhSymmetric() throws {
        let k2 = Data(hexString:
            "0000000000000000000000000000000000000000000000000000000000000002")!
        let pub1 = try ctx.pubkeyCreate(privkey: Self.KNOWN_PRIVKEY)
        let pub2 = try ctx.pubkeyCreate(privkey: k2)
        let s12 = try ctx.ecdh(privkey: Self.KNOWN_PRIVKEY, pubkey: pub2)
        let s21 = try ctx.ecdh(privkey: k2, pubkey: pub1)
        XCTAssertEqual(s12, s21, "ECDH symmetric")
    }

    func testErrorPath() {
        let zeroes = Data(count: 32)
        XCTAssertThrowsError(try ctx.pubkeyCreate(privkey: zeroes),
            "zero key should throw")
    }

    func testEcdsaDeterministic() throws {
        let sig1 = try ctx.ecdsaSign(msgHash: msg32, privkey: Self.KNOWN_PRIVKEY)
        let sig2 = try ctx.ecdsaSign(msgHash: msg32, privkey: Self.KNOWN_PRIVKEY)
        XCTAssertEqual(sig1, sig2, "RFC 6979 deterministic")
    }
}

// ── Data hex extension ──────────────────────────────────────────────────
extension Data {
    init?(hexString: String) {
        let hex = hexString.trimmingCharacters(in: .whitespacesAndNewlines)
        guard hex.count.isMultiple(of: 2) else { return nil }
        var data = Data(capacity: hex.count / 2)
        var idx = hex.startIndex
        while idx < hex.endIndex {
            let next = hex.index(idx, offsetBy: 2)
            guard let byte = UInt8(hex[idx..<next], radix: 16) else { return nil }
            data.append(byte)
            idx = next
        }
        self = data
    }
}

File: bindings/dart/test/smoke_test.dart
// ============================================================================
// UltrafastSecp256k1 — Dart Binding Smoke Test (Golden Vectors)
// ============================================================================
// Verifies dart:ffi boundary correctness using deterministic known-answer tests.
// Runs in <2 seconds.
//
// Usage:
//   dart test test/smoke_test.dart
// ============================================================================

import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:ufsecp/ufsecp.dart';

// ── Golden Vectors ──────────────────────────────────────────────────────

Uint8List hexToBytes(String hex) {
  final result = Uint8List(hex.length ~/ 2);
  for (var i = 0; i < result.length; i++) {
    result[i] = int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return result;
}

final knownPrivkey = hexToBytes(
    '0000000000000000000000000000000000000000000000000000000000000001');

final knownPubkey = hexToBytes(
    '0279BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798');

final knownXonly = hexToBytes(
    '79BE667EF9DCBBAC55A06295CE870B07029BFCDB2DCE28D959F2815B16F81798');

final sha256Empty = hexToBytes(
    'E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855');

final msg32 = Uint8List(32);
final aux32 = Uint8List(32);

// ── Tests ───────────────────────────────────────────────────────────────

void main() {
  late UfsecpContext ctx;

  setUp(() {
    ctx = UfsecpContext();
  });

  tearDown(() {
    ctx.destroy();
  });

  test('ctx_create_abi', () {
    final abi = ctx.abiVersion;
    expect(abi, greaterThanOrEqualTo(1));
  });

  test('pubkey_create_golden', () {
    final pub = ctx.pubkeyCreate(knownPrivkey);
    expect(pub, equals(knownPubkey));
  });

  test('pubkey_xonly_golden', () {
    final xonly = ctx.pubkeyXonly(knownPrivkey);
    expect(xonly, equals(knownXonly));
  });

  test('ecdsa_sign_verify', () {
    final sig = ctx.ecdsaSign(msg32, knownPrivkey);
    expect(sig.length, equals(64));
    expect(ctx.ecdsaVerify(msg32, sig, knownPubkey), isTrue);

    // Mutated → fail
    final bad = Uint8List.fromList(sig);
    bad[0] ^= 0x01;
    expect(ctx.ecdsaVerify(msg32, bad, knownPubkey), isFalse);
  });

  test('schnorr_sign_verify', () {
    final sig = ctx.schnorrSign(msg32, knownPrivkey, aux32);
    expect(sig.length, equals(64));
    expect(ctx.schnorrVerify(msg32, sig, knownXonly), isTrue);
  });

  test('ecdsa_recover', () {
    final rec = ctx.ecdsaSignRecoverable(msg32, knownPrivkey);
    expect(rec.recoveryId, inInclusiveRange(0, 3));
    final pub = ctx.ecdsaRecover(msg32, rec.signature, rec.recoveryId);
    expect(pub, equals(knownPubkey));
  });

  test('sha256_golden', () {
    final digest = UfsecpContext.sha256(Uint8List(0));
    expect(digest, equals(sha256Empty));
  });

  test('addr_p2wpkh', () {
    final addr = ctx.addrP2wpkh(knownPubkey, Network.mainnet);
    expect(addr, startsWith('bc1q'));
  });

  test('wif_roundtrip', () {
    final wif = ctx.wifEncode(knownPrivkey, true, Network.mainnet);
    final decoded = ctx.wifDecode(wif);
    expect(decoded.privkey, equals(knownPrivkey));
    expect(decoded.compressed, isTrue);
  });

  test('ecdh_symmetric', () {
    final k2 = hexToBytes(
        '0000000000000000000000000000000000000000000000000000000000000002');
    final pub1 = ctx.pubkeyCreate(knownPrivkey);
    final pub2 = ctx.pubkeyCreate(k2);
    final s12 = ctx.ecdh(knownPrivkey, pub2);
    final s21 = ctx.ecdh(k2, pub1);
    expect(s12, equals(s21));
  });

  test('error_path', () {
    final zeroes = Uint8List(32);
    expect(() => ctx.pubkeyCreate(zeroes), throwsA(isA<UfsecpException>()));
  });

  test('ecdsa_deterministic', () {
    final sig1 = ctx.ecdsaSign(msg32, knownPrivkey);
    final sig2 = ctx.ecdsaSign(msg32, knownPrivkey);
    expect(sig1, equals(sig2));
  });
}
